core/drivers/imx_rngb.c

```diff
@@ -0,0 +1,231 @@
+
+	cmd = io_read32(rng->base.va + RNGB_COMMAND);
+	cmd |= RNGB_CMD_CLR_INT | RNGB_CMD_CLR_ERR;
+	io_write32(rng->base.va + RNGB_COMMAND, cmd);
+}
+
+static void irq_unmask(struct imx_rng *rng)
+{
+	uint32_t ctrl = 0;
+
+	ctrl = io_read32(rng->base.va + RNGB_CONTROL);
+	ctrl &= ~(RNGB_CTRL_MASK_DONE | RNGB_CTRL_MASK_ERROR);
+	io_write32(rng->base.va + RNGB_CONTROL, ctrl);
+}
+
+static void rng_seed(struct imx_rng *rng)
+{
+	uint64_t tref = timeout_init_us(RNGB_TIMEOUT);
+	uint32_t cmd = 0;
+
+	cmd = io_read32(rng->base.va + RNGB_COMMAND);
+	io_write32(rng->base.va + RNGB_COMMAND, cmd | RNGB_CMD_CLR_ERR);
+	do {
+		irq_unmask(rng);
+		cmd = io_read32(rng->base.va + RNGB_COMMAND);
+		io_write32(rng->base.va + RNGB_COMMAND, cmd | RNGB_CMD_SEED);
+		wait_for_irq(rng);
+		irq_clear(rng);
+
+		if (timeout_elapsed(tref))
+			panic();
+	} while (rng->error);
+}
+
+static TEE_Result map_controller_static(void)
+{
+	if (!core_mmu_add_mapping(MEM_AREA_IO_SEC, rngb.base.pa, 0x4000))
+		return TEE_ERROR_GENERIC;
+
+	rngb.base.va = (vaddr_t)phys_to_virt(rngb.base.pa, MEM_AREA_IO_SEC);
+	if (rngb.base.va)
+		return TEE_SUCCESS;
+
+	return TEE_ERROR_GENERIC;
+}
+
+#if !defined(CFG_DT)
+static TEE_Result map_controller(void)
+{
+	return map_controller_static();
+}
+#else
+static const char *const rng_match_table[] =  {
+	"fsl,imx25-rngb",
+};
+
+static TEE_Result map_controller(void)
+{
+	void *fdt = get_dt();
+	unsigned int i = 0;
+	size_t size = 0;
+	int off = 0;
+
+	if (!fdt)
+		return TEE_ERROR_NOT_SUPPORTED;
+
+	for (i = 0; i < ARRAY_SIZE(rng_match_table); i++) {
+		off = fdt_node_offset_by_compatible(fdt, 0, rng_match_table[i]);
+		if (off >= 0)
+			break;
+	}
+
+	if (off < 0) {
+		/* rngb node not found */
+		return map_controller_static();
+	}
+
+	if (dt_enable_secure_status(fdt, off))
+		return TEE_ERROR_NOT_SUPPORTED;
+
+	if (dt_map_dev(fdt, off, &rngb.base.va, &size) < 0)
+		return TEE_ERROR_NOT_SUPPORTED;
+
+	rngb.base.pa = virt_to_phys((void *)rngb.base.va);
+
+	return TEE_SUCCESS;
+}
+#endif
+
+TEE_Result crypto_rng_read(void *buf, size_t len)
+{
+	uint32_t status = 0;
+	uint32_t words = 0;
+	uint32_t val = 0;
+	uint32_t *rngbuf = buf;
+
+	assert(buf);
+
+	if (!rngb.base.va || !rngb.base.pa)
+		return TEE_ERROR_BAD_STATE;
+
+	while (len) {
+		status = io_read32(rngb.base.va + RNGB_STATUS);
+		if (status & RNGB_STATUS_ERROR)
+			return TEE_ERROR_BAD_STATE;
+
+		words = (status & RNGB_STATUS_FIFO_LEVEL_MASK) >>
+			RNGB_STATUS_FIFO_LEVEL_SHIFT;
+
+		if (words) {
+			val = io_read32(rngb.base.va + RNGB_FIFO);
+			if (len > sizeof(uint32_t)) {
+				len = len - sizeof(uint32_t);
+				memcpy(rngbuf, &val, sizeof(uint32_t));
+				rngbuf++;
+			} else {
+				memcpy(rngbuf, &val, len);
+				len = 0;
+			}
+		}
+	}
+
+	return TEE_SUCCESS;
+}
+
+uint8_t hw_get_random_byte(void)
+{
+	uint8_t data = 0;
+
+	if (crypto_rng_read(&data, 1))
+		panic();
+
+	return data;
+}
+
+void plat_rng_init(void)
+{
+}
+
+static TEE_Result rngb_init(void)
```

Comment:
There is mentioned also version number in ID register -- should we be concerned?

Reply:
I am not aware of any erratas or quirks. if anyone is, please let me know